README.md
<p align="center">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="./assets/public/logo-dark.svg?raw=true">
    <source media="(prefers-color-scheme: light)" srcset="./assets/public/logo.svg?raw=true">
    <img alt='mion, a mikro kit for Typescript Serverless APIs' src='./assets/public/logo.svg?raw=true' width="403" height="150">
  </picture>
</p>

<p align="center">
  <strong>Benchmarks for  @mionkit/http 🚀</strong><br/>
</p>

<p align=center>
  <img src="https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square&maxAge=99999999" alt="npm"  style="max-width:100%;">
  <img src="https://img.shields.io/badge/license-MIT-97ca00.svg?style=flat-square&maxAge=99999999" alt="npm"  style="max-width:100%;">
</p>

# mion Http Benchmarks

- These benchmarks are based on the [fastify benchmarks](https://github.com/fastify/benchmarks) repo!
- `@MionKit/http` is part of the mion Framework. It uses and RPC style router!
- **This package shows how fast is mion comparatively to full featured frameworks like fastify and others.**
- You can find a full list of many other small and faster servers in the original fastify benchmarks repo.
- For metrics (cold-start) see [metrics.md](./METRICS.md)

📚 [Full mion framework documentation here!](https://github.com/MionKit/mion)

### Running & displaying the benchmarks

```sh
# start the tests
npm start

# display results in a table
npm run compare-t
```

## Other benchmarks

- **Multiple mion settings:** [MION-OPTIONS.md](MION-OPTIONS.md)  
  Running the same application but using different mion settings.
- **Cold start times:** [METRICS.md](METRICS.md)  
  Cold start times are also indicative of how the [serverless version](https://github.com/MionKit/mion/tree/master/packages/serverless) could perform in this regard, as both `@MionKit/http` an `@MionKit/serverless` are just a wrapper around `@MionKit/router` which contains all the logic.

## What's tested

The test consist of an `updateUser` request where the fields of the user must be validated, the `lastUpdate` field is a date that must be transformed into a JS Date (deserialized), and the same user must be returned back with an updated `lastUpdate` to the time of the request.

```ts
export interface User {
  id: number;
  name: string;
  surname: string;
  lastUpdate: Date;
}

// ### mion ###
// the received user by the route is already validated and deserialized
// user.lastUpdate is already a js date instead and string (result of JSON.parse)
export const routes: Routes = {
  updateUser: (context, user: User): User => {
    return {
      ...user,
      lastUpdate: new Date(),
    };
  },
};

// ### Express ###
// A plugin must be used to parse the json body
// validation must be done manually and user.lastUpdate must be deserialized manually into a date
// in this example the complexity would be in the isUser and deserializeUser functions (check src code fo that)
app.post("/updateUser", function (req, res) {
  const rawUser = req.body?.updateUser;
  if (!isUser(rawUser)) throw "app error, invalid parameter, not a user";
  const user = deserializeUser(rawUser);
  res.json({
    ...user,
    lastUpdate: new Date(),
  });
});
```

### Benchmarks

- **Machine:** darwin x64 | 8 vCPUs | 16.0GB Mem
- **Node:** `v16.18.0`
- **Run:** Thu Jul 13 2023 01:25:53 GMT+0100 (Irish Standard Time)
- **Method:** `autocannon -c 100 -d 40.03 -p 10 localhost:3000` (two rounds; one to warm-up, one to measure)

#### Req (R/s)

![benchmarks](assets/public/charts-mion/requests.png)

#### Throughput (Mb/s)

![benchmarks](assets/public/charts-mion/throughput.png)

#### Latency (ms)

![benchmarks](assets/public/charts-mion/latency.png)

#### Max Memory (Mb)

![benchmarks](assets/public/charts-mion/maxMem.png)

|           |        Version | Router |  Req (R/s)  | Latency (ms) | Output (Mb/s) | Max Memory (Mb) | Max Cpu (%) | Validation | Description                                                                                                |
| :-------- | -------------: | -----: | :---------: | -----------: | ------------: | --------------: | ----------: | :--------: | :--------------------------------------------------------------------------------------------------------- |
| http-node |        16.18.0 |      ✗ |   18978.7   |        52.18 |          4.87 |              82 |         120 |     ✗      | Super basic and completely useless bare http server, should be the theoretical upper limit in performance. |
| **mion**  |      **0.1.0** |  **✓** | **17199.2** |    **57.65** |      **4.77** |         **223** |     **143** |   **✓**    | **validation and serialization out of the box**                                                            |
| fastify   |         4.19.2 |      ✓ |   15844.0   |        62.60 |          4.08 |              87 |         118 |     -      | Validation is done using schemas and ajv. Schemas must be generated manually or using third party tools.   |
| restify   |          8.6.1 |      ✓ |   12429.4   |        79.91 |          3.21 |              94 |         116 |     ✗      | Requires third party tools.                                                                                |
| hapi      |         20.3.0 |      ✓ |   7894.6    |       126.05 |          2.02 |              96 |         126 |     ✗      | Manual validation using joi, or third party tools.                                                         |
| deepkit   | 1.0.1-alpha.75 |      ✓ |   5369.7    |       185.43 |          1.37 |             313 |         131 |     ✓      | Automatic validation out of the box (The ones that made @deepkit/types), Their rpc is way more performant. |
| express   |         4.18.2 |      ✓ |   4596.9    |       216.54 |          1.18 |             112 |         127 |     ✗      | needs third party tools, or third party tools                                                              |


HISTORY.md
# History metrics

This is the performance evolution of the mion http library across new versions.

The tricky part is to get the measurements always in the same machine to get similar results.

## Metrics

- **Machine:** darwin x64 | 8 vCPUs | 16.0GB Mem
- **Node:** `v16.18.0`
- **Run:** Sat Jul 08 2023 20:09:04 GMT+0100 (Irish Standard Time)
- **Method:** `npm run metrics` (samples: 5)
- **startup:** time elapsed to setup the application
- **listen:** time elapsed until the http server is ready to accept requests (cold start)

### Cold starts: listen time (ms) lower is better

![benchmarks](assets/public/charts/cold-starts.png)

|                                | startup(ms) | listen(ms) |
| ------------------------------ | ----------- | ---------- |
| 10-startup-deepkit-routes.js   | 1496        | 1497       |
| 10-startup-fastify-routes.js   | 159         | 237        |
| 10-startup-mion-routes.js      | 502         | 508        |
| 100-startup-deepkit-routes.js  | 1467        | 1468       |
| 100-startup-fastify-routes.js  | 179         | 381        |
| 100-startup-mion-routes.js     | 506         | 513        |
| 500-startup-deepkit-routes.js  | 1632        | 1633       |
| 500-startup-fastify-routes.js  | 263         | 836        |
| 500-startup-mion-routes.js     | 503         | 513        |
| 1000-startup-deepkit-routes.js | 1416        | 1418       |
| 1000-startup-fastify-routes.js | 425         | 1524       |
| 1000-startup-mion-routes.js    | 503         | 516        |
| 2000-startup-deepkit-routes.js | 1468        | 1469       |
| 2000-startup-fastify-routes.js | 603         | 2646       |
| 2000-startup-mion-routes.js    | 505         | 533        |
| 3000-startup-deepkit-routes.js | 1464        | 1465       |
| 3000-startup-fastify-routes.js | 874         | 3765       |
| 3000-startup-mion-routes.js    | 502         | 536        |
| 4000-startup-deepkit-routes.js | 1516        | 1518       |
| 4000-startup-fastify-routes.js | 1311        | 5212       |
| 4000-startup-mion-routes.js    | 500         | 537        |
| 5000-startup-deepkit-routes.js | 1608        | 1610       |
| 5000-startup-fastify-routes.js | 2039        | 7468       |
| 5000-startup-mion-routes.js    | 498         | 542        |
| startup-listen.js              | 537         | 539        |


lib/autocannon.js
"use strict";

const autocannon = require("autocannon");
const fs = require("fs");
const compare = require("autocannon-compare");
const path = require("path");
const { promisify } = require("util");
const pidusage = require("pidusage");
const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);
const access = promisify(fs.access);

const resultsDirectoryServer = path.join(process.cwd(), "results");
const resultsDirectoryMion = path.join(process.cwd(), "results-mion");
let resultsDirectory;

function setResultsDirectory(benchmarkName) {
  switch (benchmarkName) {
    case "mion":
      resultsDirectory = resultsDirectoryMion;
      break;
    case "servers":
    default:
      resultsDirectory = resultsDirectoryServer;
      break;
  }
}

const run = (opts = {}, handler) =>
  new Promise((resolve, reject) => {
    opts.url = "http://localhost:3000/updateUser";
    opts.method = "POST";

    // TODO: Move post param to pacakges
    opts.body =
      handler === "mion"
        ? '{"/updateUser":[{"id":123,"name":"john","surname":"smith","lastUpdate":"2020-12-17T02:24:00.000Z"}]}'
        : '{"/updateUser":{"id":123,"name":"john","surname":"smith","lastUpdate":"2020-12-17T02:24:00.000Z"}}';
    opts.headers = {
      accept: "*/*",
      "Content-Type": "application/json",
      "Content-Length": opts.body.length,
    };
    let maxMem = 0;
    let maxCpu = 0;
    let memSeries = [];
    let cpuSeries = [];
    const memCheck = setInterval(() => {
      pidusage(opts.forkedPid, (err, stats) => {
        if (err) {
          return;
        } else {
          if (stats.memory > maxMem) {
            maxMem = stats.memory;
          }
          if (stats.cpu > maxCpu) {
            maxCpu = stats.cpu;
          }
          memSeries.push(stats.memory);
          cpuSeries.push(stats.cpu);
        }
      });
    }, 1000);
    autocannon(opts, (err, result) => {
      clearInterval(memCheck);
      if (err) {
        reject(err);
      } else {
        resolve({
          ...result,
          maxMem,
          maxCpu,
          memSeries,
          cpuSeries,
        });
      }
    });
  });

const writeResult = async (handler, result) => {
  try {
    await access(resultsDirectory);
  } catch (e) {
    await mkdir(resultsDirectory);
  }

  result.server = handler;

  const dest = path.join(resultsDirectory, `${handler}.json`);
  return writeFile(dest, JSON.stringify(result));
};

module.exports.fire = async (opts, handler, save) => {
  setResultsDirectory(opts.benchmark);
  const result = await run(opts, handler);
  return save ? writeResult(handler, result) : null;
};

module.exports.compare = (a, b) => {
  const resA = require(`${resultsDirectory}/${a}.json`);
  const resB = require(`${resultsDirectory}/${b}.json`);
  const comp = compare(resA, resB);
  if (comp.equal) {
    return true;
  } else if (comp.aWins) {
    return {
      diff: comp.requests.difference,
      fastest: a,
      slowest: b,
      fastestAverage: resA.requests.average,
      slowestAverage: resB.requests.average,
    };
  }
  return {
    diff: compare(resB, resA).requests.difference,
    fastest: b,
    slowest: a,
    fastestAverage: resB.requests.average,
    slowestAverage: resA.requests.average,
  };
};
